### ISSUE TITLE
聚类算法

### ISSUE THREAD

**Issue by betterenvi** [betterenvi/EasyCon#5]
我尝试了一下 R 语言的wskm (Weighted K-means Clustering) 这个package提供的算法，包括ewkm和twkm。这两个算法跑出的结果不符合预期。

**Comment by minghuizhou** [comment]
不符合预期的原因很多啦, 您看看下述:
1, 你确定你理解了ewkm和twkm的基本工作原理?
2, 你确定你用对了ewkm和twkm?
3, 你确定你对commits量度的计算是正确的?
你在对方法完全没有描述的情况下,就claim结果不对, 一点都不靠谱.

**Comment by betterenvi** [comment]
老师说的是。我发条状态而已， 没有描述细节，很多东西还需继续探索

**Comment by betterenvi** [comment]
当前，选取如下feature：
1) 被修改文件的拓展名；
2）comment；
在过滤 **停用词 、低频词、低频拓展名以及把词换位词根**后，feature数量是几十个，依我们对”低频”定义的threshold不同而不同。比如，这是曾用过的feature：
```
xul cpp h idl html py webidl mk in build list js ipdl xhtml txt svg sjs jsm css no-ext dtd xml cachemanifest mm ini java c json gif png
merge bug build fix remove test roc tree backout 
```
确定feature之后，使用 TF*IDF 为每个commit 计算各个 feature 的值。然后使用聚类算法聚类。我尝试了 **kmeans、wskm（Weighted K-means Clustering）、NMF（Non-negative Matrix Factorization）** 等几个算法，为“低频”选用了不同的threshold， 用之前标记好的217个 commit 测试， **准确率在40%左右**。

那么：
超过多少的准确率才算还过得去呢？? 无论如何，40%的准确率不仅是没法会让自己满意的，同时也是拿不出手的。
直觉上，依靠拓展名、comment 关键词是能够帮助我们区分任务的，问题是，它们能在**多大程度**上给我们指明道路呢？ 40%？难到就是这样了吗？当然，也许我们不能对拓展名、comment 关键词抱有过高的期望，因为，它们可能只能帮我们到这了。
毕竟，在人工标记时，我们不也困难重重吗?
也许我们可以写 **个人工规则** 试试，或者 **添加** 一些规则到聚类算法中。

**Comment by minghuizhou** [comment]
1, 有尝试过使用模块作为特征吗?
2, 使用拓展名的规则是什么?
3, 人工分类的规则是什么?人工分类有尝试过使用角色来分类吗, 例如, 有些人在社区中的角色是非常明显和众所周知的(google一下那些做过很多提交的人), 或者你们跟他/或她邮件一下就可以确认他/她的角色及提交任务, 然后以这个知识来帮助benchmark是不是更靠谱一点?

我还是得说, 你这个思考的拓展度和深入度不够呀.

**Comment by betterenvi** [comment]
1. 正在把模块加入feature中；
2. 对拓展名：过滤低频的，然后使用 TF*IDF 来计算（和comment是独立计算的）
3. 人工规则，简单的例子，比如，若匹配 "backout"及变形词，归为backout类。

**Comment by betterenvi** [comment]
老师说的第三点，我似乎理解错了？
老师说的是，已有的benchmark可能不够靠谱。
我说的是，可以尝试把 规则和 聚类算法 结合一下

**Comment by minghuizhou** [comment]
对, 我说的是, 人工建立的benchmark也可能不靠谱. 因此, 我想询问: 你们人工建立benchmark的规则是啥?也就是, 你们人工靠什么来判断什么是什么?

**Comment by betterenvi** [comment]
@KyLinXYL  在 #1 这个issue中有记录，就是标记小窍门那一段

**Comment by KyLinXYL** [comment]
可以挑一些出错的样例出来看看吗？把算法给出的类别和benchmark给出的类别都附带在一起，我想看看是为什么和benchmark不符。
还有就是之前老师说有一些commit要移除掉的，可以先考虑把机器人的commit都移除了，因为我们标记的时候也搞不懂机器人在干什么，不移除可能会造成混淆

**Comment by betterenvi** [comment]
已经移出 4个 原来是unknown的commit，以及一个由hg@mozilla.org提交的commit，剩余212个。

以下是一次测试：
a. 参数设置：
number of commits to cluster: 212
extension threshold: >=1
comment word threshold: >=5
algorithm: kmeans
b. 聚类结果见：/store2/easycon/EasyCon/50.predict.ign，准确率50%
该文件的行和/store2/easycon/EasyCon/data/marked的行，一一对应。

**Comment by betterenvi** [comment]
有些commit, 如：
mozilla-central;8364:736cdbcdbe938a0f850f92aea1e78dbb64217e41;;;content/html/content/test/Makefile.in|content/html/content/test/bug392567.jar|content/html/content/test/bug392567**.jar^headers^**|content/html/content/test/test_bug392567.html;2007-11-26 22:40 -0800;dcamp;dcamp@mozilla.com;"fix test bustage from bug 369814."

在修改的文件中，**.jar^headers^** 是什么东西?

**Comment by minghuizhou** [comment]
这个可能是原始数据处理特殊符号的时候发生的问题, 把什么当成什么, 例如把回车符写成"NEW_LINE", 这里是不是"~"之类的? 可以看一下原始数据, :D. 你知道怎么看原始数据么? :dancer:

**Comment by betterenvi** [comment]
怎么看？

**Comment by minghuizhou** [comment]
http://hg.mozilla.org/mozilla-central

**Comment by betterenvi** [comment]
从数据中看到，backout的评论里面一定有backout、back out或backed out等词。
所以把comment带有backout、back out或backed out等词的commit取出，直接判定为backout类。然后对剩下的聚类，准确率约60%。在标记的212条中，算上几乎100%的backout准确率，最后准确率在66%左右。

mozilla-central共162865个commit，10097 个带有backout、back out或backed out等词。数量还是很可观的。

总结来就是，添加一些简单的、合理的规则，不仅实现上简单，在performance上也会有巨大提升。

**Comment by minghuizhou** [comment]
这个总结看上去比较靠谱. 
10097/162865 = 6.20% (一般我保留三位数字), very likely to have an impact.
还会有其他简单规则么?

**Comment by betterenvi** [comment]
最近状态，看[这里](https://github.com/betterenvi/EasyCon/commit/015699b0be220b8c3b155b4122644e6fce6d750c#commitcomment-12515251)
